### ISSUE TITLE
ENI IPAM mode should add support for assigning IP prefixes to EC2 instance

### ISSUE THREAD

**Issue by AnishShah** [cilium/cilium#16987]
## Proposal / RFE

AWS now supports assigning IP prefixes to EC2 instance. ENI IPAM mode in Cilium should also add support this as this allows increasing the pod density. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-prefix-eni.html

**Comment by lbernail** [comment]
We are very interested in this feature too!
It will help decrease the number of api calls and should make allocating IPs pretty fast.

It would be great if we could enable this on existing clusters but it will likely make the logic a bit harder (maybe we could add a prefix if no IPs have been allocated yet? so new nodes in a cluster take advantage of the feature and older ones keep using the current logic)

**Comment by webframp** [comment]
Very interested in this feature as well. EKS specific docs for using this feature are located here but assume the default Amazon VPC CNI:

* https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
* https://github.com/aws/amazon-vpc-cni-k8s/blob/master/docs/prefix-and-ip-target.md

**Comment by prune998** [comment]
Really interested in this feature too, knowing that Cilium Overlay mode does not work in EKS...
Anyone already looked at it ? Is working on a PR ? I could help or start something if nothing was done yet.

This doc is really useful too:
https://aws.amazon.com/blogs/containers/amazon-vpc-cni-increases-pods-per-node-limits/

**Comment by joestringer** [comment]
Hemanth Malla has prepared a CFP for discussion on support for this feature: https://docs.google.com/document/d/1qs6q728nud6yOr2G_wFn-h_k2DVOSo8NbNpYle4ALW8/edit?usp=sharing

**Comment by angadisachin** [comment]
We would need this feature for 2 reasons
1. Reduce cost on AWS IPAM Service (more details about the below)
2. Reduce AWS API Calls to avoid throttling on large clusters

**More details about the use-case :** Recently AWS has launched an IPAM Service (called as Amazon VPC IP Address Manager). Pricing is based on active IPs used in the account. This is a problem for users who use Amazon Elastic Kubernetes Service (EKS) or a self-managed Kubernetes cluster in AWS. These Kubernetes nodes have many IP addresses per Elastic Network Interface (ENI) and with each IP address (Pod IP) being classified as an "active IP address" to this AWS IPAM service. AWS VPC-CNI Supports prefix delegation, IPAM service counts the whole delegation (16 IP addresses) as a single IP address. We are looking for similar support with Cilium as well to reduce cost.

**Comment by aanm** [comment]
@hemanthmalla any update on this issue? I see #18463 open in draft, are you still working on it?

**Comment by hemanthmalla** [comment]
@aanm Yes. I'm waiting on e2e testing related changes to be merged from https://github.com/cilium/cilium/pull/18557 .
I can remove the dependent e2e test code and mark it as ready for review.

**Comment by ReillyBrogan** [comment]
Is there going to be a follow-up issue for adding ipv6 prefix delegation as well?

**Comment by joestringer** [comment]
@ReillyBrogan I've filed #19251 to follow up on this (cc @hemanthmalla , in case you have any inputs on that - not sure if you planned to work on this?)
